File: 0951 Flip Equivalent Binary Trees/0951.py
class Solution:
    def flipEquiv(self, a: Optional[TreeNode], b: Optional[TreeNode]) -> bool:
        if not a and not b:
            return True
        elif not a or not b or a.val != b.val:
            return False
        else:
            return (self.flipEquiv(a.left, b.left) and self.flipEquiv(a.right, b.right) 
                or self.flipEquiv(a.left, b.right) and self.flipEquiv(a.right, b.left))

File: 0951 Flip Equivalent Binary Trees/0951.cpp
class Solution {
public:
    bool flipEquiv(TreeNode* root1, TreeNode* root2) {
        if (root1 == nullptr && root2 == nullptr){
            return true;
        } else if (root1 == nullptr || root2 == nullptr || root1->val != root2->val){
            return false;
        } else {
            return (flipEquiv(root1->left, root2->left) && flipEquiv(root1->right, root2->right)) 
                || (flipEquiv(root1->left, root2->right) && flipEquiv(root1->right, root2->left));
        }
    }
};
